server/src/main/java/org/opensearch/action/admin/cluster/shards/TransportCatShardsAction.java

```diff
@@ -0,0 +1,99 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.action.admin.cluster.shards;
+
+import org.opensearch.action.admin.cluster.state.ClusterStateRequest;
+import org.opensearch.action.admin.cluster.state.ClusterStateResponse;
+import org.opensearch.action.admin.indices.stats.IndicesStatsRequest;
+import org.opensearch.action.admin.indices.stats.IndicesStatsResponse;
+import org.opensearch.action.support.ActionFilters;
+import org.opensearch.action.support.HandledTransportAction;
+import org.opensearch.action.support.TimeoutTaskCancellationUtility;
+import org.opensearch.client.node.NodeClient;
+import org.opensearch.common.inject.Inject;
+import org.opensearch.core.action.ActionListener;
+import org.opensearch.tasks.CancellableTask;
+import org.opensearch.tasks.Task;
+import org.opensearch.transport.TransportService;
+
+/**
+ * Perform cat shards action
+ *
+ * @opensearch.internal
+ */
+public class TransportCatShardsAction extends HandledTransportAction<CatShardsRequest, CatShardsResponse> {
+
+    private final NodeClient client;
+
+    @Inject
+    public TransportCatShardsAction(NodeClient client, TransportService transportService, ActionFilters actionFilters) {
+        super(CatShardsAction.NAME, transportService, actionFilters, CatShardsRequest::new);
+        this.client = client;
+    }
+
+    @Override
+    public void doExecute(Task parentTask, CatShardsRequest shardsRequest, ActionListener<CatShardsResponse> listener) {
+        final ClusterStateRequest clusterStateRequest = new ClusterStateRequest();
+        clusterStateRequest.setIsCancellationTaskRequired(true);
+        clusterStateRequest.local(shardsRequest.getLocal());
+        clusterStateRequest.clusterManagerNodeTimeout(shardsRequest.getClusterManagerNodeTimeout());
+        clusterStateRequest.clear().nodes(true).routingTable(true).indices(shardsRequest.getIndices());
+
+        clusterStateRequest.setParentTask(client.getLocalNodeId(), parentTask.getId());
+        listener = TimeoutTaskCancellationUtility.wrapWithCancellationListener(
+            client,
+            (CancellableTask) parentTask,
+            ((CancellableTask) parentTask).getCancellationTimeout(),
+            true,
+            listener
+        );
+        CatShardsResponse catShardsResponse = new CatShardsResponse();
+        ActionListener<CatShardsResponse> cancellableListener = TimeoutTaskCancellationUtility.wrapWithCancellationCheck(
```

Comment:
How is `wrapWithCancellationListener` different from `wrapWithCancellationCheck` ? 

Can `wrapWithCancellationListener` return a listener that can be used for both `cancel` and `onresponse` ? We should gurantee that the underlying restListener is not getting invoked multiple times in all the flows. 
- response after cancel
- error after cancel

Reply:
+1 why do we have 2 methods - wrapWithCancellationListener & wrapWithCancellationCheck